File: impls/nim/step1_read_print.nim
import rdstdin, types, reader, printer

proc read(str: string): MalType = str.read_str

proc eval(ast: MalType): MalType = ast

proc print(exp: MalType): string = exp.pr_str

while true:
  try:
    let line = readLineFromStdin("user> ")
    echo line.read.eval.print
  except:
    echo getCurrentExceptionMsg()

File: impls/ocaml/step1_read_print.ml
let eval ast = ast
let read str = Reader.read_str str
let print = Printer.pr_str true

let main =
  try
    while true do
      Format.printf "user> %!";
      let line = read_line () in
      try Format.printf "%a\n" print (eval (read line)) with
      | Types.MalExn exc -> Format.printf "mal exception: %a\n" print exc
      | e -> Format.printf "ocaml exception: %s\n" (Printexc.to_string e)
    done
  with End_of_file -> Format.printf "\n"
